### ISSUE TITLE
Str.regexp doesn't work well with utf-8 string

### ISSUE THREAD

**Issue by abalabokhin** [ocaml/ocaml#11269]
There some bugs related to unicode characters and regexp, eg \b doesn't work with unicode characters. When I was trying to replicate word boundary using advanced unicode characters, it doesn't work well as well. E.g, I am trying to replace "тая" or "Тая" with "тэя" or "Тэя" respectively if they are the whole words only with such a code:
```
#load "str.cma"

let regexp = Str.regexp {|\(^\|[-( \*!\?.,:;'"…«]\)\(Т\|т\)ая\($\|[-) \*!\?.,:;'"…»]\)|}
let str = "запертая ."
let s = Str.global_replace regexp {|\1\2эя\3|} str 
let _ = Printf.printf "%s" s
```
The result of the code is "запертэя .", while it should keep the string as it was "запертая .". However, if I remove the character "…" from the regexp, so it is looks like:
`{|\(^\|[-( \*!\?.,:;'"«]\)\(Т\|т\)ая\($\|[-) \*!\?.,:;'"…»]\)|}`
the code works properly, not changing the string.

Can you fix it, please?

**Comment by Octachron** [comment]
If you are looking for an unicode-aware regexp engine that works on sequence of (normalized?) extended grapheme clusters, the `Str` library is not the library that you are looking for. Moreover, the `Str` library is maintained but not actively developed. Thus I would encourage you to look at the [re2](https://github.com/janestreet/re2) or [pcre](https://mmottl.github.io/pcre-ocaml) libraries. Or, if you are concerned with human text segmentation, maybe [uuseg](https://erratique.ch/software/uuseg) would be a better fit to your need?

The `Str` library is designed to work on strings seen as sequence of 8-bits integers also known as `char` (which is not that surprising when the creation of `Str` library is concomitant with the creation of utf-8).
In other words, the `Str` library is not aware of any interpretation of char sequences as sequences of extended grapheme clusters. In particular, for `Str`, the extended grapheme cluster `…` is a sequence of three chars (`\226\128\166`). 

The result that you are seeing are thus fully within `Str` specification.

However, looking at the documentation of `Str`, it is true that this not obvious behavior is not documented.

There is thus a documentation issue at hand that I shall fix shortly, thanks for your report!

**Comment by abalabokhin** [comment]
Oh, thank you, I thought something like this, but now, with your explanation it all makes sense. My problem that I am not using ocaml directly, but a tool that uses ocaml internally (https://github.com/WeiDUorg/weidu) and support regexp with Str ocaml library. Anyway, I think I know, how to properly create word boundary for non ASCII characters using Str regexp lib now. It would be great if you fix the documentation, so other people are not confused like me.

Thank you again, I guess, you can close the issue.

**Comment by gasche** [comment]
I think that it would not be too hard to selectively/incrementally migrate weidu from Str to a unicode-aware regex engine. Looking around the codebases, there are about 500 uses of the `Str` module, which is a lot for a complete port in one go, but you could migrate the part that provoke issues for your workflow (I'm assuming that only some of the languages are running on localized text containing non-ASCII characters; it looks like Weidu embeds a small language to describe textual replacemnts to in-game dialogues), and keep the rest under `Str`. I guess you should start by opening an issue on their bug tracker.
